woocommerce/woocommerce-ios
# [Mobile Payments] Make view models responsible for subscription to connectedReaders

Fixes: #3926 
⚠️ Includes changes from #4180, to be merged first. The actual changes are in f9b1ef5

This is the last cancellable to get rid of for #3926, and possibly the most controversial 😬 

## Yosemite, Core Data, and Combine

The existing implementation had an `observeConnectedReaders` action that the store hooked into the service, calling its completion block each time the `connectedReaders` publisher sent a new value.

All the other cancellables so far were created by actual actions (connect, discover,...) that would eventually complete, but in this case, the view models want to observe just the connected reader, detached from the connect and disconnect actions. Because of this, the responsibility to cancel that subscription would fall on these view models which want to be notified. When the user navigates away from settings, they would get deallocated and stop listening for changes.

One way to shoehorn this requirement into the existing architecture would be to have the action's callback return the publisher instead, but that feels like we're bending the architectures too much.

When I look at [our documentation](https://github.com/woocommerce/woocommerce-ios/blob/develop/docs/architecture-overview.md#yosemiteframework), it has this to say about actions:

> Lightweight entities expected to contain anything required to perform a specific task.
> Usually implemented by means of Swift enums, but can be literally any type that conforms to the Action protocol.
> 
> *Allowed* to have a Closure Callback to indicate Success / Failure scenarios.
> 
> **NOTE:** Success callbacks can return data, but the "preferred" mechanism is via the EntityListener or
> ResultsController tools.

I would argue that "observing the connected readers" is not an action: it doesn't perform any task, so there's no result to send back to the caller.

However, any of the suggested alternatives (`ResultsController` or `EntityListener`) won't work, since they're both based on Core Data objects. Our architecture was clearly designed to leverage Core Data for all reads, and it doesn't accommodate any model that's not a managed object. Now we have a new `Hardware` target, where there is data that needs to flow upwards but not necessarily go through Core Data, and there's a gap in our architecture.

I believe the simplest solution for now, since there's a single use case, is for the view models to use the `ServiceLocator` to access the publishers on the `CardReaderService` directly. There is no perfect solution here, but without more than a single use case, adding extra layers of abstraction seems a bit much. If we start needing this more, maybe we can evaluate a better solution.

## Sharing mocks for testing

I was stuck for a good while trying to figure out how to share `MockCardReaderService` from the `YosemiteTests` to the `WooCommerceTests`. I didn't find any good solutions but then I saw that `MockCardReader` was already present in both targets with slightly different implementations. I'm not proud of this, but I just did the same and copied that file to the `WooCommerceTests` target. I'd love to find a better solution for sharing mocks across targets/projects, but I don't have one yet.

## To test

I used the V2 card settings with a simulated reader, and tried to hit many different configurations:

- Go to settings (should not be connected) > Connect to a reader > Disconnect > Go back
- Go to settings (should not be connected) > Connect to a reader > Go back
- Go to settings (should be connected) > Disconnect > Go back
- Go to settings (should not be connected) > Connect to a reader > Disconnect > Go back

Update release notes:
